spec/support/controller_macros.rb

```diff
@@ -0,0 +1,9 @@
+module ControllerMacros
+  def login_user
+    before(:each) do
+      @request.env["devise.mapping"] = Devise.mappings[:user]
+      user = FactoryBot.create(:user)
+      sign_in user
+    end
+  end
```

Comment:
This is _so_ slick! I really like that it keeps the declarations lean — you get to say `login_user` instead of `before { login_user }`.

If we need to know the current user within the body of the spec, does this solution provide a way to get at that object for things like `expect { post :zomg }.to change { that_user.lols.count }.by(1)` or something like that scoped to the user? It's been a while since I've written controller specs and I can never remember how they work. :-D

Also, the purpose of the macro seems to be making it declarative, but the name `login_user` has an imperative tone. Do you think it makes sense to use a declarative method name instead?

Reply:
Thanks, Jamie! I can't take credit for this solution (yay google and people wikis!!)- but do agree that it is pretty slick! :)

Also that's a solid point. I am going to adjust my code a bit and instead of calling `let(:user) { create(:user) }`, I can access the current user in the body of the spec with `subject.current_user`. Thanks for pointing that out!

In terms of making the method name more declarative, do you think something like `create_and_login_user` would be better? I'm having a bit of trouble understanding declarative vs imperative as it relates to programming, but from what I'm reading it seems to be that declarative describes *what* while imperative describes *how*?

Reply:
> I can access the current user in the body of the spec with `subject.current_user`

Oh, right! 😝 If `subject` is the controller and you use `current_user` to get the current user during a web request, it makes perfect sense that you could use the same method from the spec. Dig it!

> I'm having a bit of trouble understanding declarative vs imperative as it relates to programming, but from what I'm reading it seems to be that declarative describes _what_ while imperative describes _how_?

I think that's pretty close. To be fair the definitions of these terms are super fuzzy — it reminds me of when Alanis Morisette's song "Ironic" came out and nobody could explain what "ironic" actually meant (which I think is the entire point of the song, tbh) but a lot of people felt like the song wasn't actually describing irony. So, to me, declarative vs imperative to me is more about _when_ than what vs how. Specifically, imperative code means _now_ but declarative code might not. That's probably clear as mud so let's use an example … 😂

Rails added [`has_secure_password`](https://api.rubyonrails.org/classes/ActiveModel/SecurePassword/ClassMethods.html#method-i-has_secure_password) a while back — I would consider that a declarative name with declarative functionality. It's not imperative because it's not encrypting the password inside the class body at load time, but it sets up instances of that class to handle the encryption when you assign `password`. In an imperative form, it would probably be something like `encrypt_password`. Declarative code often wraps imperative code.

In the same way, we're not necessarily saying "log the user in right here in the block", we're saying "when this spec executes (inside the `it` block), make sure it has a logged-in user". So maybe `has_authenticated_user`?

```ruby
Rspec.Describe MyController, type: :controller do
  has_authenticated_user

  it 'idk lol' do
    expect(subject.current_user).not_to be_nil
  end
end
```

What do you think?

Reply:
Ah, I see! Thanks for that explanation, I understand now! I think `has_authenticated_user` works perfectly. I'll change it tonight. 😃